### ISSUE TITLE
Sending metrics to sensu agent

### ISSUE THREAD

**Issue by JulienChampseix** [influxdata/telegraf#97]
New features, be able to send datas output to sensu agent

**Comment by sparrc** [comment]
Thanks for the suggestion, will investigate once PR #59 is merged

**Comment by jipperinbham** [comment]
I'd be happy to work on this as I've had to integrate services with the Sensu API recently

**Comment by sparrc** [comment]
@jipperinbham that would be much appreciated! thanks!

**Comment by jipperinbham** [comment]
@JulienChampseix how were you thinking telegraf could send metrics to sensu?

**Comment by JulienChampseix** [comment]
I was thibkibg like collectd is doing. Having telegraf & sensu agent localy on the same server and teleegraf sending data to sensu which will use handler for store data into influxdb. In this case sensu will be used fir alerting part (which is currently not planned in influxdb tools stack).

**Comment by JulienChampseix** [comment]
@jipperinbham @sparrc do you have all information for start/continue this output plugin ?

**Comment by jipperinbham** [comment]
@JulienChampseix yea, I did some testing and it will not be hard to implement. The hard thing for me is I typically run these types of things in containers (docker/rkt) and the processes can't access the host level sensu client port. Not a blocker really just more of a concern.

**Comment by sparrc** [comment]
@JulienChampseix Yes but I think finding time to get this done is probably the biggest blocker ;-). At the moment kafka and statsd are bigger priorities. Sensu also presents a challenge because it's not trivial to setup like other services are (which usually just involve a `docker run` command)

As always, we happily accept pull requests :-)

**Comment by JulienChampseix** [comment]
@sparrc @jipperinbham did you find some time to work on it ?
Unfortunatly on this type of request, i could not provide PR

**Comment by JulienChampseix** [comment]
Sorry to be pushy on it but we are using sensu and having this output plugins will be a great feature

**Comment by JulienChampseix** [comment]
any update on this features/output ? @sparrc @jipperinbham

**Comment by jipperinbham** [comment]
@JulienChampseix I did some testing with this and I have a few ssues/concerns 

- the docs for writing directly to the sensu socket on a host went missing after the 0.16 release. 
- I would imagine that many use cases would involve running telegraf in a container. By doing so, it becomes quite difficult to write/access the host level socket. 
- sensu already has an extensive list of [plugins](https://github.com/sensu-plugins?tab=repositories) 

I think a potential solution would be to use the telegraf amqp output and have it sends metrics to the same RabbitMQ of your sensu system. I believe you could then wrote a handler to grab the metric and send it downstream.

**Comment by JulienChampseix** [comment]
@jipperinbham thanks for your update.

The best solution remain to send metrics to the sensu-client based on the local server. Currently that the best practise for respect the sensu architecture in term of workflow. I'm not convinced to send directly to sensu rabbitmq is a good idea and could impact alerting part (in term of management / threshold) .

My current idea for having sensu plugin for telegraf is like that:

telegraf (local) 
                      + ==> sensu-client (local)
                                                             + ==>  sensu-rabbitmq (remote) 
                                                                                                               + => sensu-uchiwa (remote-dashboard)
                                                                                                               + => manage/generate alerting

**Comment by jcmartins** [comment]
I don't know how telegraf can talk with sensu-client.
For me Telegraf is good to metrics and Sensu is good to handler alarms
If you want use Telegraf to alarm maybe the solution is:
Telegraf -> Influxdb <- Grafana v2.5 -> Grafana handler alarms.
I don't now if Chronograf have handlers to alarms, maybe can you change grafana to chronograf.

**Comment by jtyr** [comment]
Perhaps something like `threshold` plugin for Telegraf would help to solve this problem. For example Collectd has a `threshold` plugin where you can set thresholds (warnings, failure) and then the message ([OK, WARNING, FAILURE] + details) is sent via `write_sensu` plugin to Sensu client. Then there is no need to send ALL metrics to Sensu because all the threshold evaluation is done on each of the monitored machine instead of on one central place like Grafana.

**Comment by m4ce** [comment]
I believe telegraf -> influxdb -> kapacitor -> sensu is a better approach.

**Comment by jtyr** [comment]
@m4ce I don't think so. There is no support fro Sensu in kapacitor. And even if there would be, to alert on metrics stored in InfluxDB, it would require to read the metrics from the DB which puts more load on the InfluxDB servers. This is why it would be much better to generate the alerts on the client side where the metric is collected because that spreads the load across the clients instead of loading the DB with additional queries just for alerting.

**Comment by m4ce** [comment]
@jtyr, well, writing a sensu output for kapacitor should be quite trivial, but yes it is missing at the moment. One thing to keep in mind is that InfluxDB scales horizontally and kapacitor triggers do leverage InfluxDB built-in analytics functions which work in a distributed way. But yes, it could potentially lead to experience high loads, depending on the queries being executed. Definitely distributing the load across the clients is always a good approach, but that means that if you want to do some more sophisticated logic with historical values for instance, you will have to retain those on the client. I guess there are pros and cons.

**Comment by jtyr** [comment]
@m4ce I completely agree. But the user should have a choice which solution to use. So support for Sensu in Kapacitor and threshold plugin in Telegraf should be both available.

**Comment by mattyjones** [comment]
@sparrc @JulienChampseix et al.

Reading through this whole thing IMHO I think this is a dead issue or one that should be low priority. I am running telegraf -> influx -> kapacitor -> Sensu without issue. Sensu is great as a handler and Kap can generate all the alerts I personally need in a much richer environment than a sensu plugin. If anything, not to jack this, but Sensu as an input should be a higher priority. I am working on this and will open a separate issue/pr when it gets near completion.

If someone **really** needs Sensu as an output let me know and I may at least be able to help but with the current Kap functionality unless my poc goes haywire I don't see myself anyway, investing a ton of time in a Sensu output.

**Comment by sjwang90** [comment]
Closing due to lack of interest and discussion. Feel free to reopen if desired.
